### PR TITLE
Provide name for parameters in 1-arg methods when using code minings

### PR DESCRIPTION
**Pull request by jannisCode**
## What it does
This PR will enable the possability to show code minings in all Methods, also those who expect just one argument.

## Before: 
When a method only had one parameter, it was eligible for code minings, except if it was from a record class.  As you can see in the screenshot and snippet, the code minings are shown for car but not for Cat.

```java

import java.util.Objects;

class Main {
	public static void main(String...args) {
		Car c = new Car(3);
		Cat car = new Cat(5);
		car.equals(c);
		car.doSomething(12, 13);
	}
}

 class Cat {
    private int size;

    public Cat(int size) {
        this.size = size;
    }
    public Cat(long age) {}

    public int getSize() {
        return size;
    }

    public void doSomething(int a, int b) {}
    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        Cat cat = (Cat) obj;
        return size == cat.size;
    }

    @Override
    public int hashCode() {
        return Objects.hash(size);
    }

    @Override
    public String toString() {
        return "Cat{size=" + size + "}";
    }
}

public record Car(int side) {}
```

image

@mickaelistria Is there a reason, why it always checks if there are more than one parameter? Because if not, I think it would be beneficial if the code minings are always shown. 

## After this change
Code minings are always shown no matter how many Parameters there are.

## Author checklist

- [x] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

**Comment by fedejeanne** [comment]
> > Is there a reason, why it always checks if there are more than one parameter?
> 
> [...] when there is a single arg, the method name is most often sufficient to sort out what's going to happen with the given arg.

Not always, think of constructors

image

I found that having the names for methods with 1 argument is useful and also more consistent with the rest of the behavior. Thank you @jannisCode for this suggestion!

File: org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/javaeditor/codemining/CalleeJavaMethodParameterVisitor.java
Comment (fedejeanne) [inline org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/javaeditor/codemining/CalleeJavaMethodParameterVisitor.java:167]: Remove this method altogether.

**Comment by mickaelistria** [comment]
> Not always, think of constructors

The relevant case to check is when you read code. Even if there are several possible constructors, it's not likely to me that having the parameter name will help readability, eg `new Shell(SWT.ON_TOP)` or `new Shell(Display.getCurrent())` or `new Shell(mainShell)`.... read well without showing the parameter: you rarely wonder what the parameter actually is and what it's used for (unless you poorly name you variables).

> more consistent with the rest of the behavior. 

Consistency is less important that being useful and lean ;)

I would strongly hate having the parameter names in 1 method args, it would just add noise to me in the very vast majority of use cases, and useless noise hurts readability more than it helps.
So please add a preference to allow users like me to skip such parameter names annotations for those single-args methods.

**Review by mickaelistria — CHANGES_REQUESTED** [review]
Please also provide an option in preferences to skip those.

**Comment by fedejeanne** [comment]
> > more consistent with the rest of the behavior.
> 
> Consistency is less important that being useful and lean ;)

Consistency is important so that users don't end up reporting "bugs" that are not bugs and simply some "other arbitrary behavior". What I meant by that is not that the 1-arg methods don't show the name of the argument but that this behavior is **not** the same across the board: records don't exhibit this behavior.

> Please also provide an option in preferences to skip those.

I'd rather not have such an option, I think the configuration is complex enough. But there are maybe some other opinions on that. Does anyone else (other than Mickael, Jannis and me) reading this PR care to break the tie?

**Comment by mickaelistria** [comment]
> records don't exhibit this behavior.

That is more a bug to me, to fix separately.

> I'd rather not have such an option, I think the configuration is complex enough.

Please make sure you actually try to work with this proposal on various classes before considering it for a merge. My arbitrary choice was based on actual usage of the feature while it was being implemented. My impression based on that experience is that in the vast majority of cases, single arg name is noise and hurts readability.
If several people can actually try it for real, on real code, for some a few days, and report that this the hint are more often useful to them, I will stop arguing; but at the moment, I seem to be the only one who actually experimented that (some years ago) and my feedback is that this is going to hurt readability and would perceived, at least by 1 user, as a regression in usability.
So an option is a must-have to me.

File: org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/javaeditor/codemining/CalleeJavaMethodParameterVisitor.java
Comment (mickaelistria) [inline org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/javaeditor/codemining/CalleeJavaMethodParameterVisitor.java:168]: Please use a preference here
```suggestion
		return PreferenceConstants.getPreferenceStore().getBoolean(PreferencesConstants.EDITOR_JAVA_CODEMINING_SHOW_PARAMETER_NAME_SINGLE_ARG) && method.getNumberOfParameters() == 1) || method.getNumberOfParameters() == 0;
```

Comment (fedejeanne) [reply org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/javaeditor/codemining/CalleeJavaMethodParameterVisitor.java]: Rather the opposite:

```java
boolean showNamesOfSingleArguments= PreferenceConstants.getPreferenceStore().getBoolean(PreferenceConstants.EDITOR_JAVA_CODEMINING_SHOW_PARAMETER_NAME_SINGLE_ARG);
return !showNamesOfSingleArguments && method.getNumberOfParameters() == 1
		|| method.getNumberOfParameters() == 0;
```

**Comment by mickaelistria** [comment]
Thanks. That proposal is good to me!

File: org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/javaeditor/codemining/CalleeJavaMethodParameterVisitor.java
Comment (mickaelistria) [reply org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/javaeditor/codemining/CalleeJavaMethodParameterVisitor.java]: if you're into renaming with opposite meanings, I suggest renaming (and reversing conditions) from `skipParameterNamesCodeMinings` to `showParameterNamesCodeMinings() { return number > 1 || PreferenceConstants.getPreferenceStore().getBoolean(PreferencesConstants.EDITOR_JAVA_CODEMINING_SHOW_PARAMETER_NAME_SINGLE_ARG))`. That should make things simpler to understand and maintain.

File: org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/javaeditor/codemining/CalleeJavaMethodParameterVisitor.java
Comment (szarnekow) [inline org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/javaeditor/codemining/CalleeJavaMethodParameterVisitor.java:169]: Do you really want to read from the store for each method?

Comment (szarnekow) [reply org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/javaeditor/codemining/CalleeJavaMethodParameterVisitor.java:169]: And in line 179 another preference store access?

File: org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/javaeditor/codemining/CalleeJavaMethodParameterVisitor.java
Comment (fedejeanne) [reply org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/javaeditor/codemining/CalleeJavaMethodParameterVisitor.java]: No need to rename the method in this PR, it is fine as it is and it is overloaded so the renaming would have to be done for these 4 methods just so the class stays consistent:

image

File: org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/javaeditor/codemining/CalleeJavaMethodParameterVisitor.java
Comment (fedejeanne) [reply org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/javaeditor/codemining/CalleeJavaMethodParameterVisitor.java]: Would it be ok to leave that to another PR so that this one doesn't blow up?

**Comment by szarnekow** [comment]
> Would it be ok to leave that to another PR so that this one doesn't blow up?

Yes, sure.

File: org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/javaeditor/codemining/CalleeJavaMethodParameterVisitor.java
Comment (szarnekow) [reply org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/javaeditor/codemining/CalleeJavaMethodParameterVisitor.java]: Yes, sure.

**Review by fedejeanne — CHANGES_REQUESTED** [review]
I tested it and it works as expected. There is only one small typo in the label of the preferences.

After fixing that, a committer can merge this one.

File: org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/preferences/PreferencesMessages.properties
Comment (jannisCode) [reply org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/preferences/PreferencesMessages.properties]: thanks for noticing! I changed it

**Review by fedejeanne — APPROVED** [review]
The code looks sound and the minings work as expected.

Would a JDT committer be so kind to (review/test and) merge this one? Thank you in advance.

@jukzi / @jjohnstn / @iloveeclipse

**Comment by jukzi** [comment]
you  need to get a green CI first. failing `JdtTextTestSuite ParameterNamesCodeMiningTest.testRecordConstructorOneParameter` sounds related

**Comment by fedejeanne** [comment]
> you need to get a green CI first. failing `JdtTextTestSuite ParameterNamesCodeMiningTest.testRecordConstructorOneParameter` sounds related

@jukzi sorry, I missed that one.

@jannisCode, like we spoke last week, you need to replace the `testRecordConstructorOneParameter` with these 2 tests so you also test the new parameter:

```java
@Test
public void testRecordConstructorOneParameterPreferenceFalse() throws Exception {
	assertParameterNamesShown(false, 0);
}

@Test
public void testRecordConstructorOneParameterPreferenceTrue() throws Exception {
	assertParameterNamesShown(true, 1);
}

private void assertParameterNamesShown(boolean preferenceValue, int expectedShownNames) throws JavaModelException, PartInitException, InterruptedException, ExecutionException {
	PreferenceConstants.getPreferenceStore().setValue(PreferenceConstants.EDITOR_JAVA_CODEMINING_SHOW_PARAMETER_NAME_SINGLE_ARG, preferenceValue);
	String contents= """
			public class Ant {
		record Ca (int size){

		}

		Ca c = new Ca(0);

	}
	""";
	ICompilationUnit compilationUnit= fPackage.createCompilationUnit("Ant.java", contents, true, new NullProgressMonitor());
	JavaEditor editor= (JavaEditor) EditorUtility.openInEditor(compilationUnit);
	fParameterNameCodeMiningProvider.setContext(editor);
	JavaSourceViewer viewer= (JavaSourceViewer)editor.getViewer();
	waitReconciled(viewer);

	assertEquals(expectedShownNames, fParameterNameCodeMiningProvider.provideCodeMinings(viewer, new NullProgressMonitor()).get().size());
}
```
I don't have committer rights so you need to do it once you're back in the office.

**Review by fedejeanne — CHANGES_REQUESTED** [review]
Please adapt the tests

**Review by fedejeanne — COMMENTED** [review]
LGTM now. 

@jukzi ?

**Comment by jukzi** [comment]
thanks for contributing

**Comment by fedejeanne** [comment]
> thanks for contributing

Thank you for merging!

**Pull request merged**
